### PR TITLE
serveral fix for preprocessing

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

serveral fix for preprocessing

pr validation depends on #1630 

## How was this patch tested?

unit tests

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala [hunk 1]
```diff
@@ -90,7 +90,7 @@ class Graph[T: ClassTag](val inputs : Seq[ModuleNode[T]],
           .map(n => {
             n._2.fromIndex match {
               case Some(i) =>
-                if (i == 1 && n._1.element.output.isTensor) {
+                if (n._1.element.output == null || (i == 1 && n._1.element.output.isTensor)) {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: when will it be null

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BiasAdd.scala [hunk 2]
```diff
@@ -46,27 +47,52 @@ class BiasAdd[T: ClassTag]()
...
     value2d
       .addr(
         value.getTensorNumeric().one,
         onesBias,
         bias)
-
     output
   }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BiasAdd.scala]: Maybe I miss it, add unit test?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala [hunk 3]
```diff
@@ -69,34 +75,53 @@ class Pack[T: ClassTag] (val dimension: Int)(implicit ev: TensorNumeric[T])
...
     }
 
     output
   }
 
-  override def updateGradInput(input: Table, gradOutput: Tensor[_]): Table = {
-    val dimension = getPositiveDimension(input)
+  override def updateGradInput(input: Activity, gradOutput: Tensor[_]): Activity = {
+    val tableInput = input match {
+      case t: Tensor[_] => T(t)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala]: Add unit test for tensor input?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala [hunk 4]
```diff
@@ -43,10 +43,16 @@ class Pack[T: ClassTag] (val dimension: Int)(implicit ev: TensorNumeric[T])
     nDim
   }
 
-  override def updateOutput(input: Table): Tensor[_] = {
-    val dimension = getPositiveDimension(input)
+  override def updateOutput(input: Activity): Tensor[_] = {
 
-    val firstInput: Tensor[_] = input(1)
+    val tableInput = input match {
+      case t: Tensor[_] => T(t)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala]: Add unit test for tensor input?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DecodeImage.scala [hunk 5]
```diff
@@ -145,7 +145,7 @@ class DecodeGif[T: ClassTag]()(implicit ev: TensorNumeric[T])
 }
 
 class DecodeRaw[T: ClassTag](val outType: DataType,
-                             val byteOrder: ByteOrder)(implicit ev: TensorNumeric[T])
+                             val littleEndian: Boolean)(implicit ev: TensorNumeric[T])
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DecodeImage.scala]: please add a comment to explain why change it to boolean

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/NotEqual.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+import com.google.protobuf.ByteString
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+import com.intel.analytics.bigdl.tensor._
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class NotEqual[T: ClassTag]()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/NotEqual.scala]: Add unit test?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/Substr.scala [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.google.protobuf.ByteString
+import com.intel.analytics.bigdl.nn.abstractnn.Activity
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.tensor._
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.reflect.ClassTag
+
+class Substr[T: ClassTag]()
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/Substr.scala]: Add unit test?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala [hunk 8]
```diff
@@ -1162,6 +1166,36 @@ private[tensor] class DenseTensor[@specialized T: ClassTag](
 
   override def div(value: T): Tensor[T] = DenseTensorMath.mul(this, null, ev.inv(value))
 
+  override def div(x: Tensor[T]): Tensor[T] = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: add unit test?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala [hunk 9]
```diff
@@ -934,47 +927,45 @@ private[tensor] class DenseTensor[@specialized T: ClassTag](
...
     } else {
-      val targetSize = DenseTensor.expandSize(this, x)
-      val expandStrides = new Array[Int](targetSize.length)
-
-      val expandStridesX = new Array[Int](targetSize.length)
-      var i = targetSize.length - 1
-      val delta2 = targetSize.length - x.nDimension
-      while(i >= delta2) {
-        if (x.size(i + 1- delta2) != 1) expandStridesX(i) = x.stride(i + 1- delta2)
+      this.add(expandTensor(x))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: Add new unit test?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TensorflowToBigDL.scala [hunk 10]
```diff
@@ -248,7 +255,10 @@ object TensorflowToBigDL {
       FullConnectionTF, DropoutTF, Conv2D, BatchNormTF, Conv1D,
       BatchNormV2NHWCTF, BatchNormV2NCHWTF,
       FullConnectionWithoutBiasTF, Conv2D2,
-      Conv2DWithoutBias
+      Conv2DWithoutBias, NotEqualTF,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/tf/TensorflowToBigDL.scala]: Try to do not add new pattern. Move new pattern into tfloader

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BiasAdd.scala (hunk 2 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/BiasAdd.scala]: already added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala (hunk 1 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala]: when the current node is a merge, then one of its input is null

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DecodeImage.scala (hunk 5 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DecodeImage.scala]: ByteOrder is not serilizable

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala (hunk 3 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala]: added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala (hunk 4 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Pack.scala]: added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/NotEqual.scala (hunk 6 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/NotEqual.scala]: added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/Substr.scala (hunk 7 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/Substr.scala]: added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala (hunk 8 above)
Comment (yangw1234) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/DenseTensor.scala]: added

**Comment by yangw1234** [comment]
pr validation pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/TensorMath.scala [hunk 11]
```diff
@@ -411,6 +411,16 @@ trait TensorMath[T] {
   def div(value: T): Tensor[T]
 
   /**
+   * Element-wise divide
+   * x.cdiv(y) all elements of x divide all elements of y.
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/tensor/TensorMath.scala]: cdiv?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/ControlOps.scala [hunk 12]
```diff
@@ -66,14 +66,14 @@ abstract class IdentityControl[T: ClassTag]()(implicit ev: TensorNumeric[T]) ext
  * User should use ControlNodes.whileLoop or ControlNodes.switch to use this operation
  * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now
  */
-private[nn] class SwitchOps[T: ClassTag]()(implicit ev: TensorNumeric[T]) extends ControlOps[T] {
+private[bigdl] class SwitchOps[T: ClassTag]()(implicit ev: TensorNumeric[T]) extends ControlOps[T] {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/ControlOps.scala]: LGTM

**Pull request merged**